### PR TITLE
Pages: add C*UD resource, service and record logic

### PR DESCRIPTION
**Pull request by Pineirin**
Add tests
closes: https://github.com/inveniosoftware/invenio-pages/issues/70
requires: https://github.com/inveniosoftware/invenio-records-resources/pull/437

### CODE REVIEW

File: invenio_pages/services/service.py [hunk 1]
```diff
@@ -7,29 +7,117 @@
...
+            context={"identity": identity},
+            raise_errors=False,
+        )
+        page = self.record_cls.create(data)
+        response = self.result_item(
+            self, identity, page, links_tpl=self.links_item_tpl, errors=errors
+        )
+        return response
+
+    def read_url(self, identity, url):
```
Comment (ntarocco) [inline invenio_pages/services/service.py:35]: Suggestion:
```suggestion
    def read_by_url(self, identity, url):
```

File: setup.cfg [hunk 2]
```diff
@@ -50,6 +50,7 @@ tests =
     pytest-invenio>=2.1.0,<3.0.0
     invenio-db[postgresql,mysql]>=1.0.14,<2.0.0
     sphinx>=4.5.0
+    invenio-access>=1.4.5
```
Comment (ntarocco) [inline setup.cfg:53]: Was this needed in the end?

**Review by ntarocco — COMMENTED** [review]
Nice work, thank you!
I have added a few comments, it might be that some also applies to @anikachurilova changes in her PRs.

File: invenio_pages/records/models.py [hunk 3]
```diff
@@ -10,9 +10,15 @@
...
 from flask import current_app
 from invenio_db import db
+from sqlalchemy import or_
+from sqlalchemy.exc import IntegrityError
 from sqlalchemy.orm import validates
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.sql import text
 from sqlalchemy_utils.models import Timestamp
 
+from ..services.errors import PageNotFoundError
```
[the remark this answers is not in the document]
Comment (anikachurilova) [reply invenio_pages/records/models.py:20]: I think Pablo saw this pattern in my code in [invenio-banners](https://github.com/inveniosoftware/invenio-banners/blob/master/invenio_banners/records/models.py#L17). 
In my case, the identical try/except with `..NotFoundError` will be thrown in 3 places if we move it to service layer. So it looked better to just put it into the higher level.
I though agree with the dependencies order, and I do not see the reason of putting the error handler to service layer like it is now. We can move it to the records layer, WDYT @ntarocco ?

File: invenio_pages/services/service.py [hunk 4]
```diff
@@ -7,29 +7,117 @@
...
+
+    def delete(self, identity, id):
+        """Create a page."""
+        self.require_permission(identity, "delete")
+        page = self.record_cls.get(id)
+        self.record_cls.delete(page)
+        response = self.result_item(self, identity, page, links_tpl=self.links_item_tpl)
+        return response
+
+    def delete_all(self, identity):
```
[the remark this answers is not in the document]
Comment (Pineirin) [reply invenio_pages/services/service.py:98]: I agreed with the too risky thing, that's why I didn't add it as a resource.
Still, that operation was performed by invenio_app_rdm in a fixture, so I needed to add it to the service.
Of course I could instead perform a search and call the delete with every result, but from my understanding that is extremely inefficient.

File: invenio_pages/records/models.py [hunk 5]
```diff
@@ -39,23 +45,85 @@ class PageModel(db.Model, Timestamp):
...
         :returns: A :class:`invenio_pages.records.models.PageModel` instance.
         """
-        return PageModel.query.filter_by(id=id).one()
+        try:
+            return self.query.filter_by(id=id).one()
+        except NoResultFound:
+            raise PageNotFoundError(id)
+
+    @classmethod
+    def search(self, search_params={}, filters=[]):
```
Comment (kpsherva) [inline invenio_pages/records/models.py:91]: dictionary and list are mutable data collections, it is not a good practice to set them as default value of methods' arguments

File: invenio_pages/services/results.py [hunk 6]
```diff
@@ -24,23 +25,7 @@ def __init__(
...
-    @property
-    def id(self):
-        """Identity of the page."""
-        return self._record
-
-    @property
-    def links(self):
-        """Get links for this result item."""
-        return self._links_tpl.expand(self._identity, self._record)
+        super().__init__(service, identity, page, errors, links_tpl, schema)
```
Comment (kpsherva) [inline invenio_pages/services/results.py:28]: the constructor is not needed if it only calls the parent contructor

File: invenio_pages/services/service.py [hunk 7]
```diff
@@ -7,29 +7,123 @@
...
 class PageService(RecordService):
     """Page Service."""
 
-    def read_url(self, url):
-        """Retrieve a page."""
-        return Page.get_by_url(url)
+    @unit_of_work()
+    def create(self, identity, data, uow=None):
+        """Create a page."""
+        self.require_permission(identity, "create")
```
Comment (kpsherva) [inline invenio_pages/services/service.py:25]: do we have a test case to make sure only superuser(and not administrator) can create and delete?

File: invenio_pages/records/models.py [hunk 8]
```diff
@@ -39,23 +45,85 @@ class PageModel(db.Model, Timestamp):
...
+                    url=data["url"],
+                    title=data.get("title", ""),
+                    content=data.get("content", ""),
+                    description=data.get("description", ""),
+                    template_name=data["template_name"],
+                )
+                db.session.add(obj)
+
+            return obj
+        except IntegrityError:
```
Comment (ntarocco) [inline invenio_pages/records/models.py:63]: As discussed, it is probably better to raise the error when creating a page with an URL that already exists, and show to the user a nice error

File: invenio_pages/resources/resource.py [hunk 9]
```diff
@@ -24,36 +25,77 @@ class PageResource(Resource):
...
-    @request_extra_args
+    @request_data
+    @response_handler()
+    def create(self):
+        """Create a page."""
+        page = self.service.create(
+            identity=g.identity,
+            data=resource_requestctx.data or {},
+        )
+        if page.to_obj != None:
```
Comment (ntarocco) [inline invenio_pages/resources/resource.py:53]: with the raise above, you can remove this and simply return:
`return page.to_dict(), 201`

File: tests/resources/test_resources.py [hunk 10]
```diff
@@ -39,6 +45,99 @@ def test_html_content(module_scoped_pages_fixture, base_client):
...
+    page = client.post("/pages/", json=data)
+    assert page.json["title"] == "Astures"
+    assert page.status_code == 201
+
+    id = page.json["id"]
+    assert client.get(f"/pages/{id}").json["title"] == "Astures"
+
+    page = client.post("/pages/", json=data)
+    assert page.status_code == 409
+    assert page.json == {}
```
Comment (ntarocco) [inline tests/resources/test_resources.py:106]: can you add a test that a non-superuser and anonymous user cannot `create`?

File: tests/resources/test_resources.py [hunk 11]
```diff
@@ -39,6 +45,99 @@ def test_html_content(module_scoped_pages_fixture, base_client):
...
+    }
+    superuser.login(client)
+
+    page = client.post("/pages/", json=data)
+    id = page.json["id"]
+    assert client.get(f"/pages/{id}").json["title"] == "Cantabros"
+
+    client.delete(f"/pages/{id}")
+    with pytest.raises(PageNotFoundError):
+        client.get(f"/pages/{id}").json["title"] == "Cantabros"
```
Comment (ntarocco) [inline tests/resources/test_resources.py:125]: can you add a test that a non-superuser and anonymous user cannot `delete`?

File: tests/resources/test_resources.py [hunk 12]
```diff
@@ -39,6 +45,99 @@ def test_html_content(module_scoped_pages_fixture, base_client):
...
+    }
+    admin.login(client)
+
+    assert client.get("/pages/1").json["title"] != "Lusitanos"
+
+    client.put("/pages/1", json=data)
+
+    json = client.get("/pages/1").json
+    assert json["title"] == "Lusitanos"
+    assert json["template_name"] == "invenio_pages/default.html"
```
Comment (ntarocco) [inline tests/resources/test_resources.py:143]: can you add a test that a non-superuser and anonymous user cannot `update`?

**Review by ntarocco — COMMENTED** [review]
LGTM! Added a small comment to add an extra test.

**Comment by ntarocco** [comment]
As discussed, comments will be addressed in the next PR.

**Pull request merged**